lib/netutil/conn_pool.go

```diff
@@ -133,23 +145,27 @@ func (cp *ConnPool) Get() (*handshake.BufferedConn, error) {
 }
 
 func (cp *ConnPool) getConnSlow() (*handshake.BufferedConn, error) {
-	// Limit the number of concurrent dials.
-	// This should help https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2552
-	cp.concurrentDialsCh <- struct{}{}
-	defer func() {
-		<-cp.concurrentDialsCh
-	}()
-	// Make an attempt to get already established connections from the pool.
-	// It may appear there while waiting for cp.concurrentDialsCh.
-	bc, err := cp.tryGetConn()
-	if err != nil {
-		return nil, err
-	}
-	if bc != nil {
-		return bc, nil
+	for {
+		select {
+		// Limit the number of concurrent dials.
+		// This should help https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2552
+		case cp.concurrentDialsCh <- struct{}{}:
+			// Create new connection.
+			conn, err := cp.dialAndHandshake()
+			<-cp.concurrentDialsCh
+			return conn, err
+		default:
+			// Make attempt to get already established connections from the pool.
+			// It may appear there while waiting for cp.concurrentDialsCh.
+			bc, err := cp.tryGetConn()
+			if err != nil {
+				return nil, err
+			}
+			if bc != nil {
+				return bc, nil
+			}
```

Comment:
I'm not entirely sure, but I'd like to raise a question: shouldn't we add a short sleep here to avoid exhausting the CPU in a select default loop?

If concurrentDialsCh is full, the loop will continuously hit the default case, which could result in high CPU usage. I'm not sure if this is a real issue in practice, but it might be worth considering.

Reply:
I agree and saw similar case before https://github.com/VictoriaMetrics/VictoriaMetrics/blob/fcfbd6fd33d468330f69c622491845a77275bdb0/app/vminsert/netstorage/netstorage.go#L739

Added 100ms sleep before continue.